File: src/MagnusAddon/note/sentences/serialization/parsing_result_serializer.py
from __future__ import annotations

from typing import TYPE_CHECKING, override

import mylog
from autoslot import Slots
from note.notefields.json_object_field import ObjectSerializer
from note.sentences.parsed_word import ParsedMatch
from sysutils.ex_str import invisible_space, newline

if TYPE_CHECKING:
    from note.sentences.parsing_result import ParsingResult

class ParsingResultSerializer(ObjectSerializer["ParsingResult"], Slots):
    newline_replacement = f"NEWLINE{invisible_space}"
    @override
    def deserialize(self, serialized: str) -> ParsingResult:
        from note.sentences.parsing_result import ParsingResult

        rows = serialized.split(newline)
        if len(rows) < 2: return ParsingResult([], "", "")

        # noinspection PyBroadException
        try:
            return ParsingResult([ParsedMatch.serializer.from_row(row) for row in rows[2:]],
                                 self._restore_newline(rows[1]),
                                 rows[0]) if serialized else ParsingResult([], "", "")
        except Exception as ex:
            mylog.warning(f"""Failed to deserialize ParsingResult:
            message:
{ex}
{serialized}""")
            return ParsingResult([], "", "")

    def _replace_newline(self, value: str) -> str:
        return value.replace(newline, self.newline_replacement)

    def _restore_newline(self, serialized_value: str) -> str:
        return serialized_value.replace(self.newline_replacement, newline)

    @override
    def serialize(self, parsing_result: ParsingResult) -> str:
        return newline.join([parsing_result.parser_version,
                             self._replace_newline(parsing_result.sentence)]
                            + [ParsedMatch.serializer.to_row(word) for word in parsing_result.parsed_words])

File: src/src_dotnet/JAStudio.Core/Note/Sentences/Serialization/ParsingResultSerializer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Compze.Utilities.Logging;
using JAStudio.Core.Note.NoteFields;

namespace JAStudio.Core.Note.Sentences.Serialization;

public class ParsingResultSerializer : IObjectSerializer<ParsingResult>
{
   static readonly string NewlineReplacement = $"NEWLINE{StringExtensions.InvisibleSpace}";

   public ParsingResult Deserialize(string serialized)
   {
      var rows = serialized.Split('\n');
      if(rows.Length < 2)
      {
         return new ParsingResult(new List<ParsedMatch>(), "", "");
      }

      try
      {
         if(string.IsNullOrEmpty(serialized))
         {
            return new ParsingResult(new List<ParsedMatch>(), "", "");
         }

         var parsedWords = rows.Skip(2)
                               .Select(row => ParsedWordSerializer.FromRow(row))
                               .ToList();

         return new ParsingResult(
            parsedWords,
            RestoreNewline(rows[1]),
            rows[0]
         );
      }
      catch(Exception ex)
      {
         this.Log().Warning($"Failed to deserialize ParsingResult:\nmessage:\n{ex.Message}\n{serialized}");
         return new ParsingResult(new List<ParsedMatch>(), "", "");
      }
   }

   string ReplaceNewline(string value) => value.Replace("\n", NewlineReplacement);

   string RestoreNewline(string serializedValue) => serializedValue.Replace(NewlineReplacement, "\n");

   public string Serialize(ParsingResult instance)
   {
      var lines = new List<string>
                  {
                     instance.ParserVersion,
                     ReplaceNewline(instance.Sentence)
                  };

      lines.AddRange(instance.ParsedWords.Select(word => ParsedWordSerializer.ToRow(word)));

      return string.Join("\n", lines);
   }
}
